psl-core/src/main/java/org/linqs/psl/reasoner/sgd/SGDReasoner.java

```diff
@@ -60,60 +88,143 @@ public double optimize(TermStore baseTermStore) {
-        log.debug("Optimized with {} variables and {} terms.", termStore.getNumVariables(), termStore.size());
+        objective = computeObjective(termStore);
+        log.info("Optimization completed in {} iterations. Objective: {}, Normalized Objective: {}, Total Optimization Time: {}",
+                iteration, objective, objective / termCount, totalTime);
+        log.debug("Optimized with {} variables and {} terms.", termStore.getNumVariables(), termCount);
 
         termStore.syncAtoms();
 
         return objective;
     }
 
-    private boolean breakOptimization(int iteration, double objective, double oldObjective, float movement) {
+    private float minimize(SGDObjectiveTerm term, float[] variableValues, int iteration) {
+        float movement = 0.0f;
+        float variableStep = 0.0f;
+        float newValue = 0.0f;
+        int[] variableIndexes = term.getVariableIndexes();
+
+        term.computeGradient(variableValues, gradient);
+
+        for (int i = 0; i < term.size(); i++) {
+            if (coordinateStep) {
+                term.computeGradient(variableValues, gradient);
+            }
+            variableStep = computeVariableStep(variableIndexes[i], iteration);
+
+            newValue = Math.max(0.0f, Math.min(1.0f, variableValues[variableIndexes[i]] - variableStep));
+            movement += Math.abs(newValue - variableValues[variableIndexes[i]]);
+            variableValues[variableIndexes[i]] = newValue;
+        }
+
+        return movement;
+    }
+
+    private float computeVariableStep(int variableIndex, int iteration) {
+        float beta1 = 0.9f;
+        float beta2 = 0.999f;
+        float mean_hat = 0.0f;
+        float variance_hat = 0.0f;
+        float step = 0.0f;
+        float adaptedLearningRate = 0.0f;
+
+        if (adaGrad) {
+            accumulatedGradientSquares.put(variableIndex, accumulatedGradientSquares.getOrDefault(variableIndex, 0.0f)
+                    + (float)Math.pow(gradient.get(variableIndex), 2.0f));
+            adaptedLearningRate = calculateAnnealedLearningRate(iteration) / (float)Math.sqrt(accumulatedGradientSquares.get(variableIndex) + 1e-8f);
+
+            step = gradient.get(variableIndex) * adaptedLearningRate;
+        } else if (adam) {
+            accumulatedGradientMean.put(variableIndex, beta1 * accumulatedGradientMean.getOrDefault(variableIndex, 0.0f)
+                    + (1 - beta1) * gradient.get(variableIndex));
+
+            accumulatedGradientVariance.put(variableIndex, beta2 * accumulatedGradientVariance.getOrDefault(variableIndex, 0.0f) +
+                    (1 - beta2) * (float)Math.pow(gradient.get(variableIndex), 2.0f));
+
+            mean_hat = accumulatedGradientMean.get(variableIndex) / (1 - beta1);
+            variance_hat = accumulatedGradientVariance.get(variableIndex) / (1 - beta2);
+
+            adaptedLearningRate = (calculateAnnealedLearningRate(iteration) / ((float)Math.sqrt(variance_hat) + 1e-8f));
```

Comment:
I think the calculatedAnnealedLearningRate is supposed to be alpha. Assuming you are expecting to override the learningrate but i dont see it.
https://arxiv.org/pdf/1412.6980.pdf

Reply:
The annealed learning rate for these methods was intentional.

The convergence analysis for convex functions in the paper requires a decaying learning rate 
and the logisitic regression experiments use a decaying learning rate.

I think we should have options for setting the learning schedule for these methods.

Reply:
You mean it is true for both ADAM asn SGD?
if it is ok 👍 .
Just that the naming is a bit confusing though.